builtin/logical/pki/acme_state.go

```diff
@@ -516,6 +522,95 @@ func (a *acmeState) ParseRequestParams(ac *acmeContext, req *logical.Request, da
+	// Parse the key out.
+	rawProtectedBase64, ok := payload["protected"]
+	if !ok {
+		return nil, fmt.Errorf("missing required field 'protected': %w", ErrMalformed)
+	}
+	jwkBase64 := rawProtectedBase64.(string)
+
+	jwkBytes, err := base64.RawURLEncoding.DecodeString(jwkBase64)
+	if err != nil {
+		return nil, fmt.Errorf("failed to base64 parse eab 'protected': %s: %w", err, ErrMalformed)
+	}
+
+	eabJws, err := UnmarshalEabJwsJson(jwkBytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to json unmarshal eab 'protected': %w", err)
+	}
+
+	if len(eabJws.Url) == 0 {
+		return nil, fmt.Errorf("missing required parameter 'url' in eab 'protected': %w", ErrMalformed)
+	}
+	expectedUrl := ac.clusterUrl.JoinPath(expectedPath).String()
+	if expectedUrl != eabJws.Url {
+		return nil, fmt.Errorf("invalid value for 'url' in eab 'protected': got '%v' expected '%v': %w", eabJws.Url, expectedUrl, ErrUnauthorized)
+	}
+
+	if len(eabJws.Nonce) != 0 {
+		return nil, fmt.Errorf("nonce should not be provided in eab 'protected': %w", ErrMalformed)
+	}
+
+	rawPayloadBase64, ok := payload["payload"]
+	if !ok {
+		return nil, fmt.Errorf("missing required field eab 'payload': %w", ErrMalformed)
+	}
+	payloadBase64, ok := rawPayloadBase64.(string)
+	if !ok {
+		return nil, fmt.Errorf("failed to parse 'payload' field: %w", ErrMalformed)
+	}
+
+	// Make sure how eab payload matches the outer JWK key value
+	base64OuterJwk := base64.RawURLEncoding.EncodeToString(outer.Jwk)
+	if base64OuterJwk != payloadBase64 {
+		return nil, fmt.Errorf("eab payload does not match outer JWK key: %w", ErrMalformed)
+	}
+
+	rawSignatureBase64, ok := payload["signature"]
+	if !ok {
+		return nil, fmt.Errorf("missing required field 'signature': %w", ErrMalformed)
+	}
+	signatureBase64, ok := rawSignatureBase64.(string)
+	if !ok {
+		return nil, fmt.Errorf("failed to parse 'signature' field: %w", ErrMalformed)
+	}
+
+	// go-jose only seems to support compact signature encodings.
+	compactSig := fmt.Sprintf("%v.%v.%v", jwkBase64, payloadBase64, signatureBase64)
+	sig, err := jose.ParseSigned(compactSig)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing eab signature: %s: %w", err, ErrMalformed)
+	}
+
+	if len(sig.Signatures) > 1 {
+		// See RFC 8555 Section 6.2. Request Authentication:
+		//
+		// > The JWS MUST NOT have multiple signatures
+		return nil, fmt.Errorf("eab had multiple signatures: %w", ErrMalformed)
+	}
+
+	if hasValues(sig.Signatures[0].Unprotected) {
+		// See RFC 8555 Section 6.2. Request Authentication:
+		//
+		// > The JWS Unprotected Header [RFC7515] MUST NOT be used
+		return nil, fmt.Errorf("eab had unprotected headers: %w", ErrMalformed)
+	}
+
+	// Load the EAB to validate the signature against
+	eabEntry, err := a.LoadEab(ac.sc, eabJws.Kid)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to verify eab", ErrUnauthorized)
+	}
+
+	_, err = sig.Verify(eabEntry.MacKey)
```

Comment:
I'd perhaps use this payload for verification, rather than doing the verification up front. You'd want to verify the signature from the privately-held key first I think, before attempting to do something with untrusted data, especially as this is an unauthenticated endpoint. My 2c.?

Reply:
Fixed in [21300b4e9037174427fd69104c1aa3bbac37baaf](https://github.com/hashicorp/vault/commit/21300b4e9037174427fd69104c1aa3bbac37baaf)